cpp/source/deal.II/mp_values.h

```diff
@@ -41,107 +41,107 @@ class MPValues
+      typename dealii::DoFHandler<dim, spacedim>::active_cell_iterator;
 
-  virtual void
-  get_values(std::string const &key, active_cell_iterator const &cell,
-             std::vector<dealii::Tensor<1, spacedim>> &values) const;
+  CompositeMat() = default;
+
+  void get_values(std::string const &key, active_cell_iterator const &cell,
+                  std::vector<double> &values) const override;
 
 protected:
-  std::unordered_map<dealii::types::material_id, std::shared_ptr<MPValues<dim>>>
-      materials;
+  std::unordered_map<dealii::types::material_id,
+                     std::shared_ptr<MPValues<dim, spacedim>>> _materials = {};
 };
 
-//////////////////////// NEW STUFF ////////////////////////////
 template <int dim, int spacedim = dim>
-class NewStuffMPValues : public MPValues<dim, spacedim>
+class CompositePro : public MPValues<dim, spacedim>
 {
 public:
-  typedef typename dealii::DoFHandler<dim, spacedim>::active_cell_iterator
-      active_cell_iterator;
+  using active_cell_iterator =
+      typename dealii::DoFHandler<dim, spacedim>::active_cell_iterator;
 
-  NewStuffMPValues(MPValuesParameters<dim, spacedim> const &parameters);
+  CompositePro() = default;
 
-  // Needed to fix hidding of get_values.
-  using MPValues<dim, spacedim>::get_values;
+  void get_values(std::string const &key, active_cell_iterator const &cell,
+                  std::vector<double> &values) const override;
+
+protected:
+  std::unordered_map<std::string, std::shared_ptr<MPValues<dim, spacedim>>>
+      _properties = {};
+};
+
+template <int dim, int spacedim = dim>
+class UniformConstantMPValues : public MPValues<dim, spacedim>
+{
+public:
+  using active_cell_iterator =
+      typename dealii::DoFHandler<dim, spacedim>::active_cell_iterator;
+
+  UniformConstantMPValues(double const &val);
 
   void get_values(std::string const &key, active_cell_iterator const &cell,
                   std::vector<double> &values) const override;
 
 protected:
-  std::unordered_map<std::string,
-                     std::function<void(active_cell_iterator const &,
-                                        std::vector<double> &)>> properties;
+  // get_values(...) will assign _val to all elements in the vector values.
+  double _val;
 };
 
 template <int dim, int spacedim = dim>
-class PorousElectrodeMPValues : public NewStuffMPValues<dim, spacedim>
+class SuperCapacitorMPValues : public CompositeMat<dim, spacedim>
 {
 public:
-  typedef typename dealii::DoFHandler<dim, spacedim>::active_cell_iterator
-      active_cell_iterator;
-  PorousElectrodeMPValues(MPValuesParameters<dim, spacedim> const &parameters);
+  SuperCapacitorMPValues(MPValuesParameters<dim, spacedim> const &params);
+  // Use build(...) to create either an homogeneous
+  // (SuperCapacitorMPValues) or an inhomogeneous model
+  // (InhomogeneousSuperCapacitorMPValues)
+  static std::unique_ptr<MPValues<dim, spacedim>>
+  build(MPValuesParameters<dim, spacedim> const &params);
```

Comment:
That's kind of strange that build is a part of ```SuperCapacitorMPValues```. Any reason why you don't want the comment to show up in doxygen.